### PR TITLE
feat(cli): add provider upgrade command

### LINKED ISSUE #1745: Add `cdktf provider update` command
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

This was initially part of #877 but got pulled into a separate issue.

> Ideally, I would also like to be able to help me upgrade providers, e.g.
```
$> cdktf provider update hashicorp/kubernetes@latest
This project currently uses hashicorp/kubernetes 2.1.1. 2.3.2 is the latest available version. Upgrade? (yes)
// cdktf updates cdktf.json, gets it, gets the schema

Warning: the cron-job resources in hashicorp/kubernetes is in use in ./foo/bar/cron-job.ts, and the schema has changed between 2.1.1 and 2.3.2. You may need to modify this resource. Continue? (yes)
```

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
This command allows users to upgrade their local and pre-built providers
to either a specific or the latest version.

Closes #1745

### CODE REVIEW

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 1]
```diff
@@ -9,8 +9,75 @@ import { Errors } from "../errors";
 import * as fs from "fs-extra";
```
Comment (DanielMSchmidt) [inline packages/cdktf-cli/src/lib/dependencies/package-manager.ts:9]: I think this file has grown to a size where it makes sense to split each package manager into its own file. I would want to do this in a separate PR since I'm not sure which other PRs are potentially touching this and to keep the change set cleaner.

File: test/typescript/provider-upgrade-command/test.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        "^0.2.64"
+      );
+    });
+
+    it("can update withing the same cdktf version to the latest version", async () => {
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        "^0.2.55"
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random"]);
```
Comment (ansgarm) [inline test/typescript/provider-upgrade-command/test.ts:44]: Ah, nice! (If possible, I'd change the other tests to cover this case rather than the specific version one as it seems to be more common)

File: test/typescript/provider-upgrade-command/test.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      );
+
+      await driver.exec("cdktf", ["provider", "upgrade", "random@=3.2.0"]);
+
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        "^0.2.64"
+      );
+    });
+
+    it("can update withing the same cdktf version to the latest version", async () => {
```
Comment (ansgarm) [inline test/typescript/provider-upgrade-command/test.ts:39]:
```suggestion
    it("can update within the same cdktf version to the latest version", async () => {
```

File: test/typescript/provider-upgrade-command/test.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        driver.readLocalFile(".gen/versions.json")
+      );
+
+      expect(genVersionsFile["registry.terraform.io/hashicorp/random"]).toEqual(
+        "3.2.0"
+      );
+    });
+
+    onWindows(
+      "adds local provider on windows",
```
Comment (ansgarm) [inline test/typescript/provider-upgrade-command/test.ts:94]:
```suggestion
      "upgrade local provider on windows",
```

**Review by ansgarm — COMMENTED** [review]
Nice work, @DanielMSchmidt 🥳

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+class Command extends BaseCommand {
+  public readonly command = "upgrade <provider...>";
+  public readonly describe =
+    "Upgrade one or more Terraform providers in your project.";
+
+  public readonly builder = (args: yargs.Argv) =>
+    args.showHelpOnFail(true).positional("provider", {
+      type: "string",
+      desc: "Name of the provider to add. Can include a version constraint (e.g. aws@~>4.0).",
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:16]: I don't think so, it should work with a version constraint the same way add works with a version constraint 🤔

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 6]
```diff
@@ -149,6 +149,64 @@ export class DependencyManager {
...
+    console.log(`Checking if ${constraint.simplifiedName} is installed...`);
+    const packageName = await this.getPackageName(constraint);
+    logger.debug(`Expecting package ${packageName} to be installed`);
+
+    let installedPackages;
+    try {
+      installedPackages = await this.packageManager.listProviderPackages();
+    } catch (e) {
+      logger.debug(`Failed to list packages: ${e}`);
+      return undefined;
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:179]: I first thought yes, but now that you say it not being able to list is not the same as not finding a package, so I'll add a user-facing log.

[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:179]: I made it an error since it's not clear that a user would want to continue. They can retry with force local for the same outcome

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 7]
```diff
@@ -149,6 +149,64 @@ export class DependencyManager {
...
+    }
+
+    logger.debug(
+      `Installed packages: ${JSON.stringify(installedPackages, null, 2)}`
+    );
+
+    return installedPackages.find((pkg) => pkg.name === packageName)?.version;
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:186]: I don't want to accidentally catch errors outside the one spot where I expect them to be thrown

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 8]
```diff
@@ -146,15 +270,32 @@ class PythonPackageManager extends PackageManager {
...
+        .find((line) => line.includes(packageName));
+
+      logger.debug(
+        `Read requirements file and found line including ${packageName}: ${requirementLine}`
+      );
+
+      if (requirementLine) {
+        if (packageVersion ? requirementLine.includes(packageVersion) : true) {
+          console.log(
+            `Package ${packageName} already installed. Skipping installation.`
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/package-manager.ts:284]: TS has no skipping case, nor do the others. I'll just remove the output from the user-facing one

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 9]
```diff
@@ -250,6 +445,37 @@ class NugetPackageManager extends PackageManager {
...
+
+      return stdout
+        .split("\n")
+        .map((line) => {
+          // match[0] = full match
+          // match[1] = package name
+          // match[2] = a weird artifact I could not figure out how to exclude (last letter of the name)
+          // match[3] = requested version
+          // match[4] = resolved version
+          const regex = /\s*>\s((\w|\.)*)\s*(\d*\.\d*\.\d*)\s*(\d*\.\d*\.\d*)/g;
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/package-manager.ts:468]: Added an example, I don't think names help a lot, regexes are hard to read anyways :D

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts [hunk 10]
```diff
@@ -401,4 +686,48 @@ class GoPackageManager extends PackageManager {
...
+        )
+        .map((line) => {
+          const parts = line.split(" ");
+          if (parts.length !== 3) {
+            throw Errors.Internal(
+              `Expected line in go.sum to be in the format of '<package> <version> <checksum>', got: ${line}`
+            );
+          }
+
+          const name = parts[0].split("/").slice(0, 4).join("/");
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/package-manager.ts:717]: Added a comment

File: test/csharp/provider-upgrade-command/test.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+          "provider",
+          "add",
+          "random@=3.4.2", // this is not the latest version, but theres v2.0.52 of the pre-built provider resulting in exactly this package
+        ]);
+
+        expect(driver.readLocalFile("MyTerraformStack.csproj")).toContain(
+          '<PackageReference Include="HashiCorp.Cdktf.Providers.Random" Version="2.0.52" />'
+        );
+
+        await driver.exec("cdktf", ["provider", "upgrade", "random@=3.4.3"]);
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/csharp/provider-upgrade-command/test.ts:31]: The use is that I might be working around a bug in this specific version, trying to upgrade to the next patch that doesn't have the bug without making a big version jump. In this specific case the use is to have a determined specific version that we can expect in the result.

File: test/python/provider-upgrade-command/requirements.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+cdktf~=0.10.4
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/python/provider-upgrade-command/requirements.txt:1]: No, the setupPythonProject workflow sets up a pipenv project, this file is used in the pip use-case. We overwrite the cdktf.json with one that has the pip app command to simulate this. It's copied from the provider add tests

File: test/test-helper.ts [hunk 13]
```diff
@@ -115,6 +115,12 @@ export class TestDriver {
     }
   }
 
+  npmPackageVersion = (packageName: string) => {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/test-helper.ts:118]: Why? What would be a better name for the version of an npm package?

File: test/typescript/provider-add-command/test.ts [hunk 14]
```diff
@@ -42,12 +42,8 @@ describe("provider add command", () => {
...
 
-      const packageJson = JSON.parse(driver.readLocalFile("package.json"));
-
-      expect(packageJson.dependencies).toEqual(
-        expect.objectContaining({
-          "@cdktf/provider-random": expect.any(String),
-        })
+      expect(driver.npmPackageVersion("@cdktf/provider-random")).toEqual(
+        expect.any(String)
       );
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/typescript/provider-add-command/test.ts:47]: Yes, but it's harder to share. We could extend the jest matchers, but that felt like too much effort in this case.

File: test/typescript/provider-upgrade-command/test.ts (hunk 2 above)
Comment (DanielMSchmidt) [reply test/typescript/provider-upgrade-command/test.ts:44]: I think it's sufficient to test this once, the algorithm is the same across the board. I'd rather have explicit versions in the test cases since it's clearer what is happening / expected to happen

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:16]: I meant the "add" that was copied 😅 The second sentence is alright 😁

File: test/test-helper.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply test/test-helper.ts:118]: My head somehow reads it as "packageJsonVersion" 😅
As it only reads the in the dependencies of the `package.json`, maybe something like:
```suggestion
  npmPackageDependencyVersion = (packageName: string) => {
```
But that's just a nit – I'd rather like to have a different matcher (see other comment)

File: test/typescript/provider-add-command/test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply test/typescript/provider-add-command/test.ts:47]: Instead of a custom matcher we could also create a utility function that creates the matching block:
```suggestion
      expect(driver.packageJson()).toEqual(
        packageJsonWithDependency("@cdktf/provider-random")
      );
```

with this utility function
```ts
function packageJsonWithDependency(package: string, version?: string) {
  return expect.objectContaining({
    [package]: version || expect.any(String),
  });
}
```

File: test/typescript/provider-upgrade-command/test.ts (hunk 2 above)
Comment (ansgarm) [reply test/typescript/provider-upgrade-command/test.ts:44]: Yeah, makes sense for tests 👍

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/src/lib/dependencies/package-manager.ts:723]: nit: that example actually won't happen because we filter the lines above 😅

File: test/csharp/provider-upgrade-command/test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply test/csharp/provider-upgrade-command/test.ts:31]: I see the reason for using this in the test 👍 I still think the most common use-case would be to upgrade to the latest version possible.

File: packages/cdktf-cli/src/bin/cmds/handlers.ts [hunk 15]
```diff
@@ -444,3 +444,42 @@ export async function providerAdd(argv: any) {
...
+
+  const manager = new DependencyManager(
+    language,
+    cdktfVersion,
+    config.projectDirectory
+  );
+
+  let needsGet = false;
+
+  for (const provider of argv.provider) {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/bin/cmds/handlers.ts:467]: Not yet, didn't seem like the initial scope to me. Do you want to create one?

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:16]: Ah got you, the second part put me on the wrong track :D

File: packages/cdktf-cli/src/lib/dependencies/package-manager.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/package-manager.ts:723]: Yeah, but the structure is clear and I didn't have to run init and so on. Good enough :)

File: website/docs/cdktf/cli-reference/commands.mdx [hunk 16]
```diff
@@ -676,3 +676,52 @@ Add a provider from a private registry to the project.
 ```bash
 $ cdktf provider add registry.example.com/acme/my-provider
 ```
+
+## provider add
```
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:680]: can we add this to the top of the page as well? :) we've got a linked list up there. Actually, we can probably delete that now because the devportal has a nice sidecar with all the headings in it
Screen Shot 2022-10-31 at 12 24 06 PM

Comment (laurapacilio) [reply website/docs/cdktf/cli-reference/commands.mdx:680]: @DanielMSchmidt can you also adjust this page so that the commands are in alphabetical order? I feel like this one and the one before it need to be placed differently on the page to make that happen. that will make it easier for users to find what they're looking for

File: website/docs/cdktf/cli-reference/commands.mdx [hunk 17]
```diff
@@ -676,3 +676,52 @@ Add a provider from a private registry to the project.
...
+
+This command facilitates upgrading Terraform providers in a CDKTF project. If a [pre-built provider](/cdktf/concepts/providers#install-pre-built-providers) for the requested provider is installed we will update it to the newest version matching the version constraint and your `cdktf` version. Otherwise, the provider will be added to the `cdktf.json` config and `cdktf get` will be automatically invoked to generate [local provider bindings](/cdktf/concepts/providers#add-provider-to-cdktf-json).
```
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:682]:
```suggestion
This command lets you upgrade Terraform providers in a CDKTF project. If a [pre-built provider](/cdktf/concepts/providers#install-pre-built-providers) for the requested provider is installed, CDKTF updates it to the newest version that matches the version constraint and your `cdktf` version. Otherwise, CDKTF adds the provider to the `cdktf.json` configuration file and invokes cdktf get` to generate [local provider bindings](/cdktf/concepts/providers#add-provider-to-cdktf-json).
```

File: website/docs/cdktf/cli-reference/commands.mdx [hunk 18]
```diff
@@ -676,3 +676,52 @@ Add a provider from a private registry to the project.
...
+
+**Examples**
+
+Upgrade the aws provider to the newest version available for your `cdktf` version.
+
+```bash
+$ cdktf provider add aws
+```
+
+Upgrade the aws provider to the newest version available for your `cdktf` version and at version `4` of the provider version.
```
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:717]: What happens if your cdktf version doesn't support the constraint you added - like what if in this case your cdktf version doesn't support aws version 4?

File: website/docs/cdktf/cli-reference/commands.mdx [hunk 19]
```diff
@@ -676,3 +676,52 @@ Add a provider from a private registry to the project.
...
+                                  plugin cache is configured differently. Supported using the env
+                                  CDKTF_DISABLE_PLUGIN_CACHE_ENV.          [boolean] [default: false]
+      --log-level                 Which log level should be written. Only supported via setting the
+                                  env CDKTF_LOG_LEVEL                                        [string]
+  -h, --help                      Show help                                                 [boolean]
+```
+
+**Examples**
+
+Upgrade the aws provider to the newest version available for your `cdktf` version.
```
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:711]:
```suggestion
Upgrade the `aws` provider to the newest version available for your `cdktf` version.
```

File: website/docs/cdktf/cli-reference/commands.mdx (hunk 18 above)
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:717]:
```suggestion
Upgrade the `aws` provider to the newest version available for your `cdktf` version and at version `4` of the provider version.
```

File: website/docs/cdktf/cli-reference/commands.mdx [hunk 20]
```diff
@@ -676,3 +676,52 @@ Add a provider from a private registry to the project.
...
+```sh
+$ cdktf provider add --help
+```
+
+**Help Output**
+
+```
+cdktf provider upgrade <provider...>
+
+Upgrade one or more Terraform providers in your project.
```
Comment (laurapacilio) [inline website/docs/cdktf/cli-reference/commands.mdx:693]: Would it be helpful to have a similar, expanded explanation in the CLI help text as we have in docs? Just so users an understand how CDKTF handles this situation?

```
Upgrade one or more Terraform providers in your project.

If the provider has a pre-built provider already installed, CDKTF updates it to the newest version that matches the version constraint and your `cdktf` version. Otherwise, CDKTF adds the provider to the `cdktf.json` configuration file and invokes cdktf get` to generate local provider bindings.
```

**Review by laurapacilio — APPROVED** [review]
Approving because I don't want to block this work, but I had a couple of questions/suggested edits. Thank you!

File: packages/cdktf-cli/src/bin/cmds/handlers.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/cdktf-cli/src/bin/cmds/handlers.ts:467]: https://github.com/hashicorp/terraform-cdk/issues/2230

File: test/go/provider-upgrade-command/test.ts [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+      }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+      await driver.setupGoProject({
+        init: { additionalOptions: "--cdktf-version 0.13.0" },
+      });
+    });
+
+    // This test can not work since new release could have come in the mean time.
+    // We can not use an older provider version since the org move made older versions unusable.
+    // This can be set with better values once 0.14 releases.
+    test.skip("installs pre-built provider using go get", async () => {
```
Comment (ansgarm) [inline test/go/provider-upgrade-command/test.ts:22]: Is there a follow-up issue to unskip this?

File: test/test-helper.ts [hunk 22]
```diff
@@ -21,6 +21,12 @@ function execSyncLogErrors(...args: Parameters<typeof execSync>) {
   }
 }
 
+export function packageJsonWithDependency(name: string, version?: string) {
+  return expect.objectContaining({
+    [name]: version || expect.any(String),
+  });
+}
+
```
Comment (ansgarm) [inline test/test-helper.ts:29]: Thanks ❤️

**Review by ansgarm — APPROVED** [review]
(choosed the wrong radio button)

File: test/go/provider-upgrade-command/test.ts (hunk 21 above)
Comment (DanielMSchmidt) [reply test/go/provider-upgrade-command/test.ts:22]: I added it to the one for provider add: https://github.com/hashicorp/terraform-cdk/issues/1945

File: test/python/provider-upgrade-command/test.ts [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+      beforeAll(async () => {
+        driver = new TestDriver(__dirname, {
+          CDKTF_DIST: "",
+          DISABLE_VERSION_CHECK: "true",
+        }); // reset CDKTF_DIST set by run-against-dist script & disable version check as we have to use an older version of cdktf-cli
+        await driver.setupPythonProject({
+          init: { additionalOptions: "--cdktf-version 0.10.4" },
+        });
+
+        driver.copyFile("cdktf-pip.json", "cdktf.json");
```
Comment (mutahhir) [inline test/python/provider-upgrade-command/test.ts:46]: Not feedback: Ah, we don't have a way of pointing to a custom cdktf.json file? Would be useful in the future at least for testing etc. perhaps.

File: website/docs/cdktf/cli-reference/commands.mdx (hunk 16 above)
Comment (DanielMSchmidt) [reply website/docs/cdktf/cli-reference/commands.mdx:680]: Sorting them would make sense, but I'd rather do that in a second PR (or would making sub-pages be nicer?)

File: website/docs/cdktf/cli-reference/commands.mdx (hunk 18 above)
Comment (DanielMSchmidt) [reply website/docs/cdktf/cli-reference/commands.mdx:719]: CDKTF support all provider versions. If there is no pre-built provider we will generate bindings locally through cdktf get

File: test/typescript/provider-upgrade-command/test.ts [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+    });
+
+    it("can update withing the same cdktf version to the latest version in yarn", async () => {
+      // Pin random provider version so that the upgrade can do anything
+      await driver.exec("yarn", ["add", "@cdktf/provider-random@0.2.55"]);
+      expect(driver.packageJson()).toEqual(
+        packageJsonWithDependency("@cdktf/provider-random", "0.2.55")
+      );
+      await driver.exec("rm", ["-rf", "node_modules"]);
+      await driver.exec("rm", ["package-lock.json"]);
```
Comment (mutahhir) [inline test/typescript/provider-upgrade-command/test.ts:64]: Should this be `yarn.lock`? or should lines 65 & 59 also need changing from `yarn` to `npm`?

File: website/docs/cdktf/cli-reference/commands.mdx (hunk 16 above)
Comment (mutahhir) [reply website/docs/cdktf/cli-reference/commands.mdx:680]: Should this be `# provider upgrade`?

**Review by mutahhir — COMMENTED** [review]
Looks good, nice work!

Small nits here and there mostly.

File: website/docs/cdktf/cli-reference/commands.mdx (hunk 20 above)
Comment (DanielMSchmidt) [reply website/docs/cdktf/cli-reference/commands.mdx:696]: Yeah good idea :)

File: packages/cdktf-cli/src/lib/dependencies/cdktf-config-manager.ts [hunk 25]
```diff
@@ -36,4 +36,12 @@ export class CdktfConfigManager {
     currentProviders.push(provider);
     this.config.writeTerraformProviders(currentProviders);
   }
+
+  public async addProvider(constraint: ProviderConstraint): Promise<void> {
+    await this.setProvider(constraint);
+  }
+
+  public async updateProvider(constraint: ProviderConstraint): Promise<void> {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/cdktf-config-manager.ts:44]: No, I meant this to mean updateProvider as it updates the provider specs for the cdktf config. I would consider it an upgrade only after get runs which is no concern here

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 26]
```diff
@@ -149,6 +149,64 @@ export class DependencyManager {
...
   }
 
+  async upgradeProvider(constraint: ProviderConstraint) {
+    console.log(`Upgrading ${constraint.simplifiedName}...`);
+    const cdktfJson = new CdktfConfigManager();
+    const prebuiltVersion = await this.getCurrentlyInstalledVersion(constraint);
+    if (prebuiltVersion) {
+      await this.upgradePrebuiltProvider(constraint, prebuiltVersion);
+      return { addedLocalProvider: false };
+    } else if (await cdktfJson.hasProvider(constraint)) {
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:159]: Yeah good point, had it without return value first :D

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+
+  public async handleCommand(argv: any) {
+    Errors.setScope("provider upgrade");
+    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
+    const api = requireHandlers();
+    await api.providerUpgrade(argv);
+  }
+}
+
+module.exports = new Command();
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:28]: Yes, we do this for code splitting purposes. In order to have a fast code completion the initial js file needs to be small. We put all the handlers into a bigger js file with all the dependencies that is only loaded once a command is found to be running. It's not too much time it takes to load the big js, but enough to be annoying during code completion

File: packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts [hunk 28]
```diff
@@ -149,6 +149,65 @@ export class DependencyManager {
...
+        `Trying to upgrade ${constraint.simplifiedName} but it is not installed, please use "cdktf provider add ${constraint.simplifiedName}" to add it.`
+      );
+    }
+  }
+
+  async getCurrentlyInstalledVersion(constraint: ProviderConstraint) {
+    logger.info(`Checking if ${constraint.simplifiedName} is installed...`);
+    const packageName = await this.getPackageName(constraint);
+    logger.debug(
+      `Expecting package ${packageName} to be installed if provider is installed as pre-built one`
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/lib/dependencies/dependency-manager.ts:173]: `getPackageName` returns the name of the package, it does not check which packages are installed. the listProviderPackages command lists the installed packages, not knowing which packages should be installed. I found this heplful during debugging to see which package we are expecting and which we are detecting. if there is a bug in either implementation we can easily find it this way

File: test/python/provider-upgrade-command/cdktf.json [hunk 29]
```diff
@@ -0,0 +1,10 @@
+{
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply test/python/provider-upgrade-command/cdktf.json:1]: I kept this in line with the provider add test, we can do a quick follow up to change both, but I don't want to do it here since this PR keeps on dragging for too long already

File: test/python/provider-upgrade-command/test.ts (hunk 23 above)
Comment (DanielMSchmidt) [reply test/python/provider-upgrade-command/test.ts:46]: Yeah we heard this sporadically from customers as well. Having multiple makes it a bit harder to think about what a cdktf project is. With the current implementation it's "a directory containing a cdktf.json" and I think that simplicity is okay for now.

File: test/typescript/provider-upgrade-command/test.ts (hunk 24 above)
Comment (DanielMSchmidt) [reply test/typescript/provider-upgrade-command/test.ts:64]: No, `package-lock.json` is correct. The normal project setup installs dependencies through npm, we are cleaning up all npm artifacts and install with yarn to simulate a yarn project

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (mutahhir) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:28]: Hmm, I wonder if this will actually code split though? I don't know much about it, but I thought the only code split that happens is with the [dynamic imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import). Doesn't matter for this PR though, was just wondering.

**Review by mutahhir — APPROVED** [review]

File: packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/cdktf-cli/src/bin/cmds/provider-upgrade.ts:28]: The dynamic import is done here: https://github.com/hashicorp/terraform-cdk/blob/main/packages/cdktf-cli/src/bin/cmds/helper/utilities.ts#L33
